File: java/algoDS/swapNoTempBitwise/TestSwapNoTempBitwise.java
public class TestSwapNoTempBitwise {
    public static void main(String[] args) {
	int a = 2;
	int b = 3;
	System.out.println("[V] TestSwapNoTempBitwise: input  a=[" + a + "], b=[" + b + "]");
	swapNoTempBitwise(a, b);
        System.out.println("[V] TestSwapNoTempBitwise: output a=[" + a + "], b=[" + b + "]");

    }

    public static void swapNoTempBitwise(int a, int b) {
	System.out.println("[D] swapNoTempBitwise: input   a=[" + a + "], b=[" + b + "]");

	a = a ^ b;
        System.out.println("[V] swapNoTempBitwise: step 1  a=[" + a + "], b=[" + b + "]");

	b = a ^ b;
        System.out.println("[V] swapNoTempBitwise: step 2  a=[" + a + "], b=[" + b + "]");

	a = a ^ b;
	System.out.println("[V] swapNoTempBitwise: step 3  a=[" + a + "], b=[" + b + "]");

	// Note: For Java, int is primitive, so it is passed by value, and output will not change !!!
    }
}

File: cpp/algoDS/swapNoTempBitwise/TestSwapNoTempBitwise.cpp
#include <stdio.h>

void swapNoTempBitwise(int* a, int* b);

int main() {
    int a = 2;
    int b = 3;
    printf("[V] TestSwapNoTempBitwise: before: a=[%d], b=[%d]\n", a, b);
    swapNoTempBitwise(&a, &b);
    printf("[V] TestSwapNoTempBitwise: after : a=[%d], b=[%d]\n", a, b);
    
    return 0;
}

void swapNoTempBitwise(int* a, int* b) {
    printf("[D] swapNoTempBitwise: input *a=[%d], *b=[%d]\n", *a, *b);
    if (a == NULL || b == NULL) {
	printf("[E] swapNoTempBitwise: a or b is NULL !!!");
	return;
    }

    *a = *a ^ *b;
    printf("[V] swapNoTempBitwise: step 1 *a=[%d], *b=[%d]\n", *a, *b);

    *b = *a ^ *b;
    printf("[V] swapNoTempBitwise: step 2 *a=[%d], *b=[%d]\n", *a, *b);

    *a = *a ^ *b;
    printf("[V] swapNoTempBitwise: step 3 *a=[%d], *b=[%d]\n", *a, *b);
}

